### PR TITLE
snap-confine: validate SNAP_NAME against security tag

### PR DESCRIPTION
**Pull request by stolowski**
Make sure SNAP_NAME env variable matches the name from security tag.

### CODE REVIEW

File: cmd/libsnap-confine-private/snap.c [hunk 1]
```diff
@@ -27,25 +27,31 @@
...
+	    "^snap\\.([a-z](-?[a-z0-9])*)\\.([a-zA-Z0-9](-?[a-zA-Z0-9])*|hook\\.[a-z](-?[a-z])*)$";
 	regex_t re;
-	if (regcomp(&re, whitelist_re, REG_EXTENDED | REG_NOSUB) != 0)
+	if (regcomp(&re, whitelist_re, REG_EXTENDED) != 0)
 		die("can not compile regex %s", whitelist_re);
 
-	int status = regexec(&re, security_tag, 0, NULL, 0);
+  // first capture is the entire string, second is the name we care about
+  regmatch_t matches[2];
+	int status = regexec(&re, security_tag, sizeof(matches)/sizeof(regmatch_t), matches, 0);
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.c]: `sizeof matches / sizeof *matches`

File: cmd/libsnap-confine-private/snap.c [hunk 2]
```diff
@@ -27,25 +27,31 @@
...
 	regfree(&re);
 
-	return (status == 0);
+  if (status != 0 || matches[1].rm_so < 0) {
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.c]: Can you please document that this checks the sub-match associated for snap name was actually found.

File: cmd/libsnap-confine-private/snap.c [hunk 3]
```diff
@@ -27,25 +27,34 @@
 #include "string-utils.h"
 #include "cleanup-funcs.h"
 
-bool verify_security_tag(const char *security_tag)
+bool verify_security_tag(const char *security_tag, const char *snap_name)
 {
 	// The executable name is of form:
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap.c]: Can you move this to the documentation of the function in the header file please

File: cmd/libsnap-confine-private/snap-test.c [hunk 4]
```diff
@@ -23,42 +23,46 @@
...
-	g_assert_false(verify_security_tag("SNAP.name.app"));
-	g_assert_false(verify_security_tag("snap.Name.app"));
-	g_assert_false(verify_security_tag("snap.0name.app"));
-	g_assert_false(verify_security_tag("snap.-name.app"));
-	g_assert_false(verify_security_tag("snap.name.@app"));
-	g_assert_false(verify_security_tag(".name.app"));
-	g_assert_false(verify_security_tag("snap..name.app"));
-	g_assert_false(verify_security_tag("snap.name..app"));
-	g_assert_false(verify_security_tag("snap.name.app.."));
+	g_assert_false(verify_security_tag
```
Comment (zyga) [inline cmd/libsnap-confine-private/snap-test.c]: Can you move this test to the end of the function and add a comment to explaining that it snap name and security tag are not matching (though are separately valid).

Comment (stolowski) [inline cmd/libsnap-confine-private/snap-test.c]: Ok, done.

**Comment by stolowski** [comment]
@jdstrand Thanks for the review, I've fixed the comments per your suggestions.

**Pull request merged**